CMakeLists.txt

```diff
@@ -269,37 +287,41 @@ if(R_LANG)
   endif(BUILD_TESTING)
 endif(R_LANG)
 
+if(PYTHON OR MATLAB OR R_LANG)
+  # Disable shared library and demo exe when building the interfaces
+  set(OSQP_BUILD_SHARED_LIB OFF)
+  set(OSQP_BUILD_DEMO_EXE OFF)
+endif()
+
 # ----------------------------------------------
 # Static Library - osqpstatic
 # ----------------------------------------------
-function(OSQP_BUILD_STATIC_LIB)
+if( OSQP_BUILD_STATIC_LIB )
   add_library(osqpstatic STATIC $<TARGET_OBJECTS:OSQPLIB>)
   set_target_properties(osqpstatic PROPERTIES OUTPUT_NAME osqp)
   if(ALGEBRA_DEFAULT)
     # Transitive dependency on OBJECT library does not work. See https://gitlab.kitware.com/cmake/cmake/-/issues/18682
     target_sources(osqpstatic PRIVATE $<TARGET_OBJECTS:qdldlobject>)
   endif()
   target_include_directories(
-    osqpstatic PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    osqpstatic PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/public;${CMAKE_CURRENT_BINARY_DIR}/include/public>"
                       "$<INSTALL_INTERFACE:$<INSTALL_PREFIX>/${CMAKE_INSTALL_INCLUDEDIR}/osqp>")
 
   if(MATLAB)
     target_link_libraries(osqpstatic ${Matlab_LIBRARIES})
   endif(MATLAB)
-  set(OSQP_STATIC_LIB
-      ON
-      PARENT_SCOPE)
-endfunction()
+endif()
 
 # ----------------------------------------------
 # Shared Library - osqp
 # ----------------------------------------------
-function(OSQP_BUILD_SHARED_LIB)
-  add_library(osqp SHARED $<TARGET_OBJECTS:OSQPLIB>)
```

Comment:
Is there an advantage to adding the *sources* to the shared library (which also means we have to target-include-directories `osqplib_includes` to make it work) instead of the compiled objects from OSQPLIB as we were doing before? I'm sure there's a good reason, just curious for my own understanding :)

Reply:
This comes from marking the public API with `OSQP_API` to get the default visibility from the shared library. When the object files were included here, all our public functions were still with hidden visibility even though they had the `OSQP_API` annotation. The only way to fix that was to make the shared library compile the object files again (this time with the correct build flag to make the public API visible). Technically, we only need to compile the functions that have their definitions inside the public headers again to get the right visibility information in the library, but I found it simplest to just compile everything instead of having to sort out which source files had public/private functions (we probably only need to recompile about 7 functions instead of the entire codebase).

Reply:
Thanks @imciner2 - makes sense. I see the `BUILDING_OSQP` definition and how you're using it here. Very instructive!